packages/app/src/cli/services/environment/id-matching.ts

```diff
@@ -1,26 +1,24 @@
 import {ExtensionRegistration} from '../dev/create-extension.js'
 import {IdentifiersExtensions} from '../../models/app/identifiers.js'
 import {Extension} from '../../models/app/extensions.js'
+import {Result, err, ok} from '@shopify/cli-kit/common/result'
 
-export type MatchResult =
-  | {
-      result: 'ok'
-      identifiers: IdentifiersExtensions
-      pendingConfirmation: {extension: Extension; registration: ExtensionRegistration}[]
-      toCreate: Extension[]
-      toManualMatch: {local: Extension[]; remote: ExtensionRegistration[]}
-    }
-  | {
-      result: 'invalid-environment'
-    }
+export interface MatchResult {
+  identifiers: IdentifiersExtensions
+  pendingConfirmation: {extension: Extension; registration: ExtensionRegistration}[]
+  toCreate: Extension[]
+  toManualMatch: {local: Extension[]; remote: ExtensionRegistration[]}
+}
 
 export async function automaticMatchmaking(
   localExtensions: Extension[],
   remoteRegistrations: ExtensionRegistration[],
   identifiers: {[localIdentifier: string]: string},
-): Promise<MatchResult> {
+): Promise<Result<MatchResult, Error>> {
```

Comment:
Regarding your comments:
* We could add also a ResultAsync<T, E> with T being a Promise as a wrapper of Promise<Result<T, E>>
* I don't think we should migrate all the code for using this new type. I think it is suitable for methods like this that makes you to design en structure that could return a correct response or an error. I guess that any other person that has to design a solution will go with another different design
* In case of using third party libraries like here, we should wrap the code to catch the error and return a custom `Result` 
https://github.com/Shopify/cli/blob/9f461c0adeb5b47564ff4ec9dd6b1ee7bfc732b4/packages/cli-kit/src/api/common.ts#L71:L94